add hyprland profile

odomingao opened the pull request.

AveraheLinuxUser commented:
Hello, every gui app needs acess to read /dev/shm/@{uuid} as owner, which created by hyprland after launch.
It can be add as new abstraction called hyprland, or added to exists desktop abstraction

/etc/apparmor.d/abstractions/hyprland

# apparmor.d - Full set of apparmor profiles
# Copyright (C) 2022-2024 Alexandre Pujol <alexandre@pujol.io>
# SPDX-License-Identifier: GPL-2.0-only

  owner /dev/shm/@{uuid} r,

# vim:syntax=apparmor

odomingao commented:
> Hello, allmost every gui app needs acess to read /dev/shm/@{uuid} as owner, which created by hyprland after launch. It can be add as new abstraction called hyprland, or added to exists desktop abstraction
> 
> /etc/apparmor.d/abstractions/hyprland
> 
> owner /dev/shm/@{uuid} r,

Yes, the desktop abstraction probably makes the most sense but there are graphical applications that don't need the abstraction. I think it should be just `/dev/shm/@{uuid} r,` rather than `owner /dev/shm/@{uuid} r,`, because graphical applications can be run as other users depending on each individual's setup.

roddhjav commented:
`owner /dev/shm/@{uuid} r,` us required for wayland, thefore it needs to be added to the wayland abstraction, not in the desktop abs.

roddhjav commented on apparmor.d/groups/hyprland/hyprland:
```diff
@@ -0,0 +1,71 @@
+  include <abstractions/base>
+  include <abstractions/desktop>
+  include <abstractions/graphics>
+  include <abstractions/wayland>
```
the wayland abs is already included in the desktop abs. Therefore, you can safely remove it here.

roddhjav commented on apparmor.d/groups/hyprland/hyprland:
```diff
@@ -0,0 +1,71 @@
+
+  @{exec_path} mr,
+
+  @{bin}/** rPUx,
```
Use the app-launcher-user abstraction instead of this line.

roddhjav commented on apparmor.d/groups/hyprland/hyprland:
```diff
@@ -0,0 +1,71 @@
+  /usr/share/hyprland/{,*} r,
+  /usr/share/libinput/{,*} r,
+
+  owner @{user_cache_dirs}/hyprland/** w,
```
It is safe, and time saving to directly allow `rw` on its own cache.

roddhjav commented on apparmor.d/groups/hyprland/hyprland:
```diff
@@ -0,0 +1,71 @@
+  /usr/share/libinput/{,*} r,
+
+  owner @{user_cache_dirs}/hyprland/** w,
+  owner @{user_config_dirs}/hypr/hyprland.conf r,
```
Same for config, it is it own config file:
```
owner @{user_config_dirs}/hypr/** r,
```

roddhjav commented on apparmor.d/groups/hyprland/hyprland:
```diff
@@ -0,0 +1,71 @@
+  owner @{user_cache_dirs}/hyprland/** w,
+  owner @{user_config_dirs}/hypr/hyprland.conf r,
+
+  @{run}/systemd/sessions/@{int} r,
+  @{run}/udev/data/+acpi:* r,
+  @{run}/udev/data/+dmi:id r,
+  @{run}/udev/data/+drm:card@{int}-* r,
+  @{run}/udev/data/+input:input@{int} r,
+  @{run}/udev/data/+pci:* r,
+  @{run}/udev/data/+platform:pcspkr r,
+  @{run}/udev/data/+sound:card@{int} r,
+  @{run}/udev/data/+usb:* r,
+  @{run}/udev/data/c13:@{int} r,
+  @{run}/udev/data/c189:@{int} r,
+  @{run}/udev/data/c226:@{int} r,
```
It is easier to keep (and copy paste) existing structure already used around the profiles:

```suggestion
  @{run}/systemd/sessions/@{int} r,
 
  @{run}/udev/data/+acpi:* r,             # for acpi
  @{run}/udev/data/+dmi:id r,             # for motherboard info
  @{run}/udev/data/+drm:card@{int}-* r,   # For screen outputs
  @{run}/udev/data/+input:input@{int} r,  # for mouse, keyboard, touchpad
  @{run}/udev/data/+pci:* r,              # Identifies all PCI devices (CPU, GPU, Network, Disks, USB, etc.)
  @{run}/udev/data/+platform:* r,
  @{run}/udev/data/+sound:card@{int} r,   # for sound card
  @{run}/udev/data/+usb* r,               # for USB mouse and keyboard
  @{run}/udev/data/c13:@{int} r,          # for /dev/input/*
  @{run}/udev/data/c189:@{int}  r,        # for /dev/bus/usb/**
  @{run}/udev/data/c226:@{int} r,         # for /dev/dri/card* 
```

Also, please move all `@{run}/udev/data` rules after the temp directories as they are quite different (`@{run}/user/@{uid}`, `/tmp/`, `/dev/shm`)

roddhjav commented on apparmor.d/groups/hyprland/hyprland:
```diff
@@ -0,0 +1,71 @@
+  owner @{run}/user/@{uid}/.hyprpaper_* rw,
+  owner @{run}/user/@{uid}/.hyprpicker_* rw,
+  owner /tmp/.X@{int}-lock w,
+  owner /dev/shm/@{uuid} rw,
```
Move this to the wayland complete file.

odomingao replied:
is it ok to remove `owner` from  `owner /dev/shm/@{uuid} rw,`? Applications running as a different user will not launch otherwise

roddhjav replied:
I would say no, (simply add a rule without owner in you local addition). This kind of runtime specific settings does not fit well with generic apparmor profiles. This kind would better be handled directly by the program that handle starting the app as a different user.

roddhjav commented on apparmor.d/groups/hyprland/hyprland:
```diff
@@ -0,0 +1,71 @@
+
+  @{sys}/bus/ r,
+  @{sys}/class/input/ r,
+  @{sys}/devices/@{pci}/oot_vga r,
```
oot_vga does not exist. It should be `@{sys}/devices/@{pci}/boot_vga r,`

odomingao commented:
I think that I fixed everything you mentioned. This: `owner /dev/shm/grim-@{rand6} rw,` is needed for XDPH. I added it here because I thought it makes sense to be in the wayland abstraction, but it's not needed for the Hyprland profile itself

roddhjav commented:
Thanks. It looks good now. Merged
